src/librustc_mir/build/block.rs

```diff
@@ -79,4 +80,31 @@ impl<'a,'tcx> Builder<'a,'tcx> {
             block.unit()
         })
     }
+
+    // Helper method for generating MIR inside a conditional block.
+    pub fn with_cond<F>(&mut self, block: BasicBlock, span: Span,
+                        cond: Operand<'tcx>, f: F) -> BasicBlock
+    where F: FnOnce(&mut Builder<'a, 'tcx>, BasicBlock) -> BasicBlock {
+        let scope_id = self.innermost_scope_id();
+
+        let then_block = self.cfg.start_new_block();
+        let else_block = self.cfg.start_new_block();
+
+        self.cfg.terminate(block, scope_id, span,
+                           TerminatorKind::If {
+                               cond: cond,
+                               targets: (then_block, else_block)
+                           });
+
+        let after = f(self, then_block);
+
+        // If the returned block isn't terminated, add a branch to the "else"
+        // block
```

Comment:
What is the purpose of this hack? Does it always result in correct code in case of bugs? If not, then lack of the terminator should just stay a bug, as it currently is.